[ARM] enable FENV_ACCESS pragma support for hard-float targets

Varnike opened the pull request:
Added support for FENV_ACCESS pragma on hard-float ARM platforms. Also changes were made to clang/test/Parser/pragma-fp-warn.c so that for thumbv7a only the soft-float-abi target case is checked.

davemgreen commented:
I believe the backend would still need work to make sure this is supported, which has not been done yet. I was expecting it to fail more noisily, but it appears the strict nodes are lowered to generic nodes. That doesn't mean that strict-fp is supported by the Arm backend, as it might not handle the instructions correctly at the moment (moving instructions past one another, for example).

davemgreen commented:
@john-brawn-arm did the work for AArch64, I'm not sure if he would have an idea how much work it would involve for the Arm backend.

john-brawn-arm commented:
It's been several years since I looked at this, but from my notes what's needed before setting HasStrictFP=true is:

- Instruction selection patterns for strict fp ops. This is probably just changing e.g. "fadd" to "any_fadd" as I expect for most or all of the instruction selection patterns work fine with strict fp.
- Instructions that can raise fp exceptions need to have mayRaiseFPException=1. Without this transformations can move around or just remove instructions that raise fp exceptions.
- We need proper modelling of reads/write of the FPSCR rounding control bits for controllable rounding modes to be handled correctly.

john-brawn-arm commented:
Simple example where setting HasStrictFP=true without doing the above gives wrong results:
```
int fetestexcept( int excepts );
#define FE_DIVBYZERO 0x04

int fn(float x) {
#pragma STDC FENV_ACCESS ON
  x / 0;
  return fetestexcept(FE_DIVBYZERO);
}
```
With --target=aarch64-none-elf we get a divide instruction and so return FE_DIVBYZERO, with --target=arm-non-eabi we get no divide instruction and return 0.

Varnike committed:
[ARM] Enable strict fp support

Varnike commented:
Hello!

Thank you for the comments and sorry for the delayed response.

I've now returned to this task and plan to work towards full support for strict fp on hard-float ARM targets.

Based on @john-brawn-arm's example and suggestions, I was able to add support for STRICT_FDIV. With such changes the example is compiled correctly (with divide instruction). Could you please review these changes and let me know if you spot any issues? Are there any additional cases that require special handling? Also, is there a test suite that would be sufficient to cover most strict fp usage scenarios? 

If these changes are acceptable, I plan to proceed with support for the remaining strict fp ops.

Varnike commented:
Now I have an approximate volume of work and plan to try to do it. I have provided these edits rather as an example of correction for a particular case (not taking tests into account), by analogy with which other cases can be considered. Therefore, I would like to know how correct and complete they are.

davemgreen commented:
@john-brawn-arm any ideas? It looks OK to me. (It would need to handle all instructions before HasStrictFP was added though).

Varnike commented:
@john-brawn-arm, @davemgreen can you please take a look at the patch?

davemgreen reviewed (commented):
I believe it looks sensible. We will need strict-fp tests for the new operations now supported, and there is quite a bit going on in this patch. Neon and MVE sometimes don't implement ieee-fp exactly, I would need to remind myself of the details.

I would not expect the tests to change (like llvm/test/CodeGen/Thumb2/mve-fmas.ll) - strictfp should not make the normal case worse. Do you know what might be causing those to be different?

Varnike commented:
Fix test failures reported by CI.

davemgreen reviewed (commented):
It looks like this is progressing nicely. It's quite a large change and if there are any parts we can pull out and commit separately, that would help.

Varnike commented on llvm/lib/Target/ARM/ARMRegisterInfo.td:
```diff
@@ -186,9 +186,7 @@ def APSR       : ARMReg<15, "apsr">;
 def SPSR       : ARMReg<2,  "spsr">;
 def FPSCR      : ARMReg<3,  "fpscr">;
-def FPSCR_NZCV : ARMReg<3,  "fpscr_nzcv"> {
-  let Aliases = [FPSCR];
```
Yes, in general they should be aliases. However, this behavior will lead to the test changes discussed above (e.g. Thumb2/mve-fmas.ll): they occur because some of the instructions begin to change the FPSCR (that is, FPSCR_NZCV too because they are aliases), so during scheduling it is no longer possible to leave everything as is.

> I think we need to teach it that the fpscr bits of nofpexcept instructions and fpscr_nzcv do not interact and can be treated separately for scheduling. 

My understanding is that while they alias, this approach won’t work (please correct me if I'm wrong). To keep the existing tests unchanged, it might be necessary to change the patterns for most instructions which can raise fp exception and change FPSCR:  instead of using any_* add individual patterns fot strict_*. This can lead to significant complications in writing patterns for fp. 

 Can you tell if you see a simpler way to solve this? Also how critical is the fact that some tests (e.g. Thumb2/mve-fmas.ll) may change?

Varnike commented:
- Added remaining tests for f16/32/64 intrinsic.
- Fixed lowering for several f16 ops, which leaded to errors at selection stage (https://godbolt.org/z/G8E9Ksv4P).
- The only untested intrinsics are maximum and minimum. At the moment they also lead to compilation fail (https://godbolt.org/z/cbdx5bj9v, same for llvm.maximum and minimum). Is there support for these intrinsics in the backend and should it be addressed within this patch?

I haven’t addressed FPSCR yet; before proceeding I'd like to hear your opinion on how should it be done.

davemgreen commented:
Am I right that the bits of FPSCR that are read by the instruction are just the rounding mode? Can we add a new register for just those bits, like FPSCR_NZCV. FPSCR_RM would alias with FPSCR but not FPSCR_NZCV.
```
def FPSCR_RM   : ARMReg<3,  "fpscr_rm"> {
  let Aliases = [FPSCR];
}
```
I didn't try it past a few of the basic tests in llvm. The important part is that we don't hoist the instructions that use the rounding mode past an instructions that sets it. (And an instruction that sets FPSCR doesn't get moved past one that sets FPSCR_NZCV).

The alternative might be to try and make fpscr_rm and fpscr_nzcv SubRegs of fpscr, which might work but ran into other issues when I tried it.

> As for how to organize the patches after separation: should each individual patch have a separate PR?

Ideally yeah, So long as they are individually testable. It is more difficult to review larger changes and get them to stick in llvm without causing issues and needing to be reverted.

Varnike commented:
> Can we add a new register for just those bits, like FPSCR_NZCV. FPSCR_RM would alias with FPSCR but not FPSCR_NZCV.

Thanks for the comment! I took this approach and it seems that everything works with it.

I managed to split the patch into 4 parts that can be tested independently:

- #160696

- #160698

- #160699

- commit 06f4aea from this pr which enables strict fp support by setting HasStrictFP = true. It can only be applied after the previous patches have been committed. Would it be preferable to leave this as a single commit in the current PR, or to move it into a separate new one?

However, this split leaves three tests, added by analogy with AArch64, that I haven’t yet included in any patch. One of them ([strict-fp-ops.ll](https://github.com/llvm/llvm-project/blob/700169d93e593001731d7e2b92a60c09e161f520/llvm/test/CodeGen/ARM/strict-fp-ops.ll)) depends on both patches, which I still wanted to keep separate for review convenience. The other two ([strict-fp-int-promote.ll](https://github.com/llvm/llvm-project/blob/700169d93e593001731d7e2b92a60c09e161f520/llvm/test/CodeGen/ARM/strict-fp-int-promote.ll), [strictfp_f16_abi_promote.ll](https://github.com/llvm/llvm-project/blob/700169d93e593001731d7e2b92a60c09e161f520/llvm/test/CodeGen/ARM/strictfp_f16_abi_promote.ll)) are already supported correctly without my changes. Perhaps it would be better to add them after the other patches have been committed to LLVM? If this approach is not appropriate and the tests should be included in the patches above, please let me know and I will update them.

Varnike commented:
Once the necessary strict fp support has been added to the arm backend, HasStrictFP=true can be enabled.

Varnike commented:
> Once the necessary strict fp support has been added to the arm backend, HasStrictFP=true can be enabled.

@davemgreen, could this be submitted?

davemgreen commented:
@john-brawn-arm do you know what else is needed to support strict-fp?

I think we probably need to do something about vector operations - both MVE and Neon. (Which might be annoying if we need to add new intrinsics for all of them as they do not match the expected behaviour of the fp instructions).

john-brawn-arm commented:
> @john-brawn-arm do you know what else is needed to support strict-fp?
> 
> I think we probably need to do something about vector operations - both MVE and Neon. (Which might be annoying if we need to add new intrinsics for all of them as they do not match the expected behaviour of the fp instructions).

Looking at what happens if I do ``llc llvm/test/CodeGen/AArch64/fp-intrinsics-vector.ll -mtriple=arm-none-eabi -mcpu=cortex-a55``, the strict vector operations are legalized into a sequence of strict scalar operations, which are then selected as normal. This is happening because we don't have any specific handling of the strict vector operations in the ARM target, and this is the default behaviour.

I _think_ this is is the correct thing to do, as both NEON and MVE instructions ignore the FPSCR exception and rounding control bits and treat them as if they were zero, so to get the correct strict behaviour we need to use scalar instructions.

I think we do need to have a test of this i.e. have a version of the fp-intrinsics-vector.ll test in llvm/test/CodeGen/ARM.

Varnike commented:
ping

Varnike commented:
No, should I add it?

davemgreen commented:
We might need it to stop the automatic mutation of strict-fp nodes to non-strict equivalents. All the targets that support strict-fp have it set to true, so I assume we should be doing the same. It looks like some of the existing tests fail with it on.

davemgreen reviewed (commented):
> If necessary, this can be moved into a separate PR.

Yeah that might be useful - so we can add this part before turning the whole feature on. There are some strict-fp vector operations that could theoretically be supported - I'm not sure yet if they are very worth adding or not.

davemgreen commented on llvm/test/CodeGen/ARM/fp-intrinsics-vector.ll:
```diff
@@ -0,0 +1,1571 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 6
+; RUN: llc -mtriple=armv7a-- -mattr=+neon,+vfp4 %s -o - | FileCheck %s
```
Making this armv7a-none-eablhf can help make the tests a little cleaner.

Varnike replied:
changed, but it didn't affect anything

davemgreen replied:
Sorry - that was a typo and should have been `armv7a-none-eabihf`.

Varnike commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -614,16 +614,25 @@ ARMTargetLowering::ARMTargetLowering(const TargetMachine &TM_,
         setOperationAction(Op, MVT::f64, Legal);
+      
+      setOperationAction(ISD::STRICT_FP_ROUND, MVT::f32, Legal);
+      setOperationAction(ISD::STRICT_FP_EXTEND, MVT::f64, Legal);
```
1. Corrected for  `fpext f16->f64` case

according to checks in `LowerFP_EXTEND`:
```
  assert((!Subtarget->hasFP64() || !Subtarget->hasFPARMv8Base()) &&
         "With both FP DP and 16, any FP conversion is legal!");
```
if both features are enabled we can mark it Legal and mark Custom otherwise.

3. For `f128->f32` case `fptrunc` will be softened anyway, since f128 is not legal.

davemgreen commented:
I put together these for the various MVE intrinsics
https://github.com/llvm/llvm-project/pull/169156
https://github.com/llvm/llvm-project/pull/169771
https://github.com/llvm/llvm-project/pull/169798
https://github.com/llvm/llvm-project/pull/169797
https://github.com/llvm/llvm-project/pull/169795

There is one last one for fprtosi and sitofp that I still need to make (it was being difficult as it involves a type parameter). The same is probably needed for Neon too, to prevent the intrinsics from scalarizing. If you can remove the clang change from this commit (or create a new one) I think it looks OK.

Varnike commented:
Moved changes to #170136. This PR has been reverted to its original state.

Varnike commented:
> On systems without fullfp16, fp16 nodes do not lower at the moment https://godbolt.org/z/3qh3fbWjG. (For systems without fp64 we do manage to produce libcalls).

Addressed this in #173666. 

> There are a number of problems with the various min/max's, some that are present for both standard nodes and strictfp.

Could you please provide more details about the issue?
